src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,900 @@
+
+        #region SPI transfer
+
+        /// <summary>
+        /// Conduct a SPI transaction
+        /// </summary>
+        /// <param name="dataOut">A byte array to send.The length of the array will determine how many bytes are transferred.</param>
+        /// <returns>Returns an array of the bytes read.</returns>
+        public byte[] SpiTransferArray(byte[] dataOut)
+        {
+            byte[] result = new byte[dataOut.Length];
+            _spiDevice.TransferFullDuplex(dataOut, result);
+            return result;
+        }
+
+        /// <summary>
+        /// Read a 32 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Return the value read</returns>
+        public int SpiRead32(SpiMessageType MessageType)
+        {
+            int retVal = -1;
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = (int)(reply[4] << 24) | (reply[5] << 16) | (reply[6] << 8) | reply[7];
+            }
+            else
+            {
+                throw new IOException($"{nameof(SpiRead32)} : no SPI response");
+            }
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read a 16 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Return the vallue read</returns>
+        public int SpiRead16(SpiMessageType MessageType)
+        {
+            int retVal = -1;
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0, 0, };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == 0xA5)
+            {
+                retVal = (int)(reply[4] << 8) | reply[5];
+            }
+            else
+            {
+                throw new IOException($"{nameof(SpiRead16)} : no SPI response");
+            }
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Read a 8 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Return the vallue read</returns>
+        public int SpiRead8(SpiMessageType MessageType)
+        {
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+            Console.WriteLine($"{reply[0]} {reply[1]} {reply[2]} {reply[3]} {reply[4]}");
+            return (reply[3] == 0xA5) ? reply[4] : throw new IOException($"{nameof(SpiRead16)} : no SPI response");
+        }
+
+        /// <summary>
+        /// Send a 8 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <param name="Value">The value to be sent</param>
+        public void SpiWrite8(SpiMessageType MessageType, int Value)
```

Comment:
why not directly take byte as an argument?

Reply:
adjusted and for the Spi16 as well